zebra/if_netlink.c

```diff
@@ -403,64 +600,135 @@ netlink_interface (struct sockaddr_nl *snl, struct nlmsghdr *h,
          slave_kind = RTA_DATA(linkinfo[IFLA_INFO_SLAVE_KIND]);
 #endif
 
-      if (kind && strcmp(kind, "vrf") == 0)
-        {
-          vrf_device = 1;
-          netlink_vrf_change(h, tb[IFLA_LINKINFO], name);
-          vrf_id = (vrf_id_t)ifi->ifi_index;
-        }
+      netlink_determine_zebra_iftype (kind, &zif_type);
+    }
+
+  /* If VRF, create the VRF structure itself. */
+  if (zif_type == ZEBRA_IF_VRF)
+    {
+      netlink_vrf_change(h, tb[IFLA_LINKINFO], name);
+      vrf_id = (vrf_id_t)ifi->ifi_index;
     }
 
   if (tb[IFLA_MASTER])
     {
       if (slave_kind && (strcmp(slave_kind, "vrf") == 0))
-        vrf_id = *(u_int32_t *)RTA_DATA(tb[IFLA_MASTER]);
+        {
+          zif_slave_type = ZEBRA_IF_SLAVE_VRF;
+          vrf_id = *(u_int32_t *)RTA_DATA(tb[IFLA_MASTER]);
+        }
+      else if (slave_kind && (strcmp(slave_kind, "bridge") == 0))
```

Comment:
This only works from Linux 3.14. Before that, no `IFLA_INFO_SLAVE_KIND` is attached to an enslaved interface. Unfortunately, it's quite difficult to know if the interface is enslaved to a bridge or a bond at this point. Moreover, during bootstrap, we can't check if the upper index is one from a bridge because we may not have this information yet. Other than doing an additional netlink request (not pretty), I don't see an easy way out of this without adding many lines of code (looking for an already known bridge or storing the upper link index then scanning "unknown" slaves after discovering a bridge).

Otherwise, just document that EVPN only works with kernel 3.14+.

There is similar code in `netlink_link_change()`.